### PR TITLE
feat(google-genai): add util-backed 2.x instrumentation

### PR DESCRIPTION
**Pull request by sipercai**
# Description

This PR adds LoongSuite instrumentation for the Google Gen AI Python SDK 2.x, backed by `opentelemetry-util-genai`.

The provider implementation is aligned with the canonical OpenTelemetry Python GenAI Google instrumentation (1.0b1/current upstream snapshot), while retaining LoongSuite's shared GenAI util for completion hooks, content capture, multimodal extensions, and client metrics.

Supported synchronous and asynchronous surfaces include:

- `models.generate_content`
- `models.generate_content_stream`
- `models.embed_content`
- `interactions.create`, including streaming
- automatic Python function tools

The instrumentation records LLM, EMBEDDING, and TOOL spans; request/response metadata; response IDs; finish reasons; token usage; captured messages; streaming TTFT; embedding dimensions; errors; and GenAI client metrics.

After PR #245, instrumentation-only work is isolated with `hook_advice`: provider/application calls and stream iteration execute exactly once outside advice, while telemetry prepare, mapping, chunk, close/finalize, and wrapper-construction failures degrade telemetry without changing the provider result, chunk identity/order, cancellation, `GeneratorExit`, or original exception.

Google ADK changes and Robin's commercial `_SUPPRESS_LLM_SDK_KEY` overlay are intentionally out of scope.

Fixes # (N/A)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Python 3.9–3.13, oldest/latest Google GenAI dependency matrix
- [x] `205 passed, 95 skipped` on the latest package environment
- [x] Dedicated fail-open tests for prepare/mapping/finalizer/reporter/chunk/wrapper/close/aclose/cancellation/GeneratorExit/cross-context/concurrency/tool failures
- [x] Fresh-cache `tox -e precommit`
- [x] `tox -e generate-workflows`
- [x] Package wheel and sdist build
- [x] Sanitized cassette replay with `--vcr-record=none`
- [x] Bounded live `google-genai 2.14.0` Gemini verification with CMS readback

The live matrix covered sync/async non-streaming, sync/async streaming, embeddings, interactions, interaction streaming, deterministic function tools, multi-turn calls, two concurrent clients, early close/aclose, a real provider error, and injected response/chunk telemetry failures. CMS read back all 20 scenario traces and all 42 exported spans, including LLM → TOOL, EMBEDDING, TTFT, token usage, and ERROR trees.

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR:
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### Spec and Scope

- Direct feature implementation scoped to the Google Gen AI SDK provider plugin.
- Canonical upstream baseline and delta are documented in `UPSTREAM.md`.
- No Google ADK or Robin commercial behavior is included.

### Local Checks

| Check | Result |
| --- | --- |
| Static readiness | pass |
| Python 3.9–3.13 oldest/latest tox matrix | pass |
| Latest focused tests | 205 passed, 95 skipped |
| Package Ruff lint | pass |
| Fresh-cache precommit | pass |
| Workflow generation | pass, generated files stable |
| Wheel and sdist build | pass |
| Cassette privacy scan | pass, zero credential/path matches |

### Real E2E Matrix

| Scenario | Result | Observed telemetry |
| --- | --- | --- |
| sync/async generation | pass | LLM response/model/token attributes |
| sync/async SSE | pass | ordered chunks, TTFT, token usage |
| sync/async embeddings | pass | EMBEDDING spans and dimensions |
| sync/async interactions | pass | `interactions.create` LLM spans |
| deterministic function tool | pass | LLM parent with TOOL child |
| multi-turn | pass | two sibling LLM spans under one scenario |
| two concurrent clients | pass | isolated trace/context state |
| early close/aclose | pass | span finalized once |
| invalid provider model | pass | original `ClientError`, ERROR telemetry |
| response-mapping advice failure | pass | original response preserved, telemetry safely degraded |
| every-chunk advice failure | pass | all provider chunks preserved, telemetry safely degraded |

### Upstream comparison

- [OpenTelemetry Python GenAI Google instrumentation](https://github.com/open-telemetry/opentelemetry-python-genai/tree/main/instrumentation/opentelemetry-instrumentation-google-genai)
- [OpenInference Google GenAI instrumentation](https://github.com/Arize-ai/openinference/tree/main/python/instrumentation/openinference-instrumentation-google-genai)

Both upstream implementations contain useful local span-isolation patterns, but still mix some request/stream telemetry callbacks with the business boundary. The LoongSuite delta keeps the upstream provider mapping while applying the stricter PR #245 fail-open lifecycle.

### CODE REVIEW

File: instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/generate_content.py [hunk 1]
```diff
@@ -0,0 +1,1021 @@
+# Copyright The OpenTelemetry Authors
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/generate_content.py:1]: **[Info]** This file is 1021 lines — consider splitting the `_GenerateContentHandler` (span/metric lifecycle) from the pure mapping helpers (`_map_*`, `_extract_*`) into separate modules. The mapping functions are stateless and testable in isolation; separating them would reduce cognitive load when navigating the instrumentation lifecycle.

File: instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/_compat.py [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+"""
+
+from __future__ import annotations
+
+import logging
+import os
+import timeit
+from dataclasses import asdict, dataclass
+from types import TracebackType
+from typing import Any
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/_compat.py]: **[Info]** The bridge pattern here is well-documented — the module docstring clearly explains this is a removable impedance match. One suggestion: consider adding a `# TODO(loongsuite-util-migration): remove after util migrates to 1.x` marker at the module level so the cleanup target is greppable.

File: instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/_compat.py [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+            self.attributes[_TOOL_DEFINITIONS] = gen_ai_json_dumps(
+                [asdict(tool) for tool in self.tool_definitions]
+            )
+        if _captures_content_on_span():
+            message_attributes = (
+                (_INPUT_MESSAGES, self.input_messages),
+                (_OUTPUT_MESSAGES, self.output_messages),
+                (_SYSTEM_INSTRUCTIONS, self.system_instruction),
+            )
+            for key, messages in message_attributes:
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/_compat.py:218]: **[Warning]** In `_prepare_finish`, `self.output_tokens = (self.output_tokens or 0) + self.thinking_tokens` silently adds reasoning tokens to the total output token count. This is correct if the upstream semantic convention defines `output_tokens` as inclusive of reasoning, but if it should be exclusive, this would inflate the reported count. Worth confirming against the GenAI semantic conventions spec — or adding a comment clarifying the intended behavior.

File: instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/_stream.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    normal iteration. The subclass should pass the SDK stream to
+    ``super().__init__(stream)`` and implement the three telemetry hooks:
+    ``_process_chunk`` for per-chunk state, ``_on_stream_end`` for successful
+    finalization, and ``_on_stream_error`` for failure finalization.
+
+    Users should consume subclasses as normal streams, for example with
+    ``for chunk in wrapper`` or ``with wrapper``. The hook methods are called
+    internally by the wrapper lifecycle and are not part of the public API.
+    """
+
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/_stream.py:99]: **[Info]** The `_StreamWrapperMeta` metaclass resolving both `ABCMeta` and `wrapt.ObjectProxy`'s metaclass is a clean solution. The comment explaining why this is needed is helpful for future maintainers.

File: instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/generate_content.py [hunk 5]
```diff
@@ -0,0 +1,1021 @@
...
+from .dict_util import flatten_dict
+from .message import (
+    to_input_messages,
+    to_output_messages,
+    to_system_instructions,
+)
+from .tool_call_wrapper import wrapped_tool
+
+_is_mcp_imported = False
+McpClientSession = McpTool = None
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/generate_content.py:70]: **[Info]** The MCP client import guard (`_is_mcp_imported`) is correctly placed at module level with `try/except ImportError`. One minor note: the global `_is_mcp_imported` flag is set to `False` initially and then potentially overwritten. If this module is imported before `mcp` is installed but `mcp` is installed later at runtime, the flag won't update. This is fine for the typical use case (dependencies are fixed at install time), but worth being aware of in dynamic environments.

**Review by ralf0131 — COMMENTED** [review]
### Summary
This PR adds comprehensive Google GenAI Python SDK 2.x instrumentation backed by `opentelemetry-util-genai`. The implementation follows the established LoongSuite instrumentation patterns, supports both sync/async surfaces (generate_content, embed_content, interactions), and includes thorough fail-open error handling via the `hook_advice` mechanism. Test coverage is extensive (205 passed, 95 skipped) with dedicated fail-open tests for every telemetry failure mode.

### Findings
- **[Info]** `generate_content.py:1` — 1021-line file could benefit from splitting mapping helpers into a separate module
- **[Info]** `_compat.py:31` — Bridge module is well-documented; consider a greppable TODO marker for the cleanup target
- **[Warning]** `_compat.py:218` — Reasoning tokens silently added to output_tokens total; verify this matches the GenAI semantic convention intent
- **[Info]** `_stream.py:97` — `_StreamWrapperMeta` metaclass pattern is clean and well-commented
- **[Info]** `generate_content.py:69` — MCP import guard is correct for typical install-time dependency resolution

### Suggestions
1. **File size**: Consider extracting `_map_generate_content_config`, `_extract_*` helpers and the `_MessageConverter` into a `_mapping.py` module. This would make `generate_content.py` focus purely on the instrumentation lifecycle.
2. **Reasoning token accounting**: Add a brief comment or assertion clarifying whether `output_tokens` should include or exclude reasoning tokens per the semantic conventions.
3. **UPSTREAM.md**: The upstream alignment documentation is excellent. Consider keeping it updated as the upstream 1.0b1 evolves.

### Overall
Solid, well-tested instrumentation that aligns with both upstream OTel conventions and LoongSuite's existing patterns. The fail-open design via `hook_advice` is consistently applied across all surfaces. No blocking issues — the findings above are informational.

---
*Automated review by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
Reviewed the core instrumentation structure. The implementation follows established OTel instrumentation patterns with clean separation of concerns:

**Architecture** ✅
- `instrumentor.py` follows `BaseInstrumentor` contract correctly
- `_compat.py` bridge layer properly adapts upstream 1.0 API to LoongSuite extensions
- `_stream.py` handles sync/async streams with proper context manager protocol and finalization
- `generate_content.py` covers sync/async generate_content, embed_content, and interactions

**Key Observations**
- AllowList utility supports regex patterns with proper escaping
- Stream wrappers use `wrapt.ObjectProxy` with correct metaclass for type compatibility
- Token usage accounting and TTFT tracking are properly integrated
- Test coverage is comprehensive with VCR cassettes for deterministic replay

**Minor Note**
- Large PR (10K+ lines) — maintainer may want to verify upstream semantic convention alignment separately

Overall the code is well-structured and follows the project's established patterns.

---
*Automated review by github-manager-bot*

**Comment by ralf0131** [comment]
⚠️ **Merge conflict detected**

This PR has conflicts with the `main` branch and cannot be merged. Please rebase or merge `main` into your branch and resolve the conflicts:

```bash
git fetch origin
git checkout feat/google-genai-observability
git rebase origin/main
# resolve conflicts, then:
git push --force-with-lease
```

This is a one-time reminder. Feel free to @mention me for a re-review after conflicts are resolved.

---
*Automated notification by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
### Summary
Comprehensive Google GenAI SDK 2.x instrumentation package aligned with upstream OTel `opentelemetry-instrumentation-google-genai==1.0b1`. Well-structured with proper fail-open semantics, sync/async parity, streaming TTFT support, embeddings, interactions, and automatic function tool wrapping. Test coverage (205 passed, 95 skipped) is thorough including fail-open edge cases.

### Findings
- **[Info]** `generate_content.py` is 1021 lines — consider splitting `_instrument_generate_content` and `_uninstrument_generate_content` into a separate module if it grows further, to keep the file navigable.
- **[Info]** The `_is_mcp_imported` module-level flag with `McpClientSession`/`McpTool` globals is fine for optional dependency handling, but worth noting that importing `mcp` at module load time could trigger side effects in environments where `mcp` is installed but not intended for use with this instrumentor.

### Suggestions
- The UPSTREAM.md baseline documentation is excellent. Consider adding a brief note on the divergence points (e.g., `hook_advice` integration, multimodal extensions) to help future maintainers track what's LoongSuite-specific vs upstream.

---
*Automated review by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
### Summary
Re-review after new commits since last approval (July 31). Latest changes include streaming output message aggregation fix and additional test coverage.

### Key Changes Since Last Review
- **Streaming output aggregation** (`generate_content.py`, `message.py`): Introduces `StreamOutputMessageAccumulator` to properly aggregate streaming deltas by candidate index. This ensures `gen_ai.output.messages` contains complete logical responses instead of one message per SSE chunk — a correct fix for a real data quality issue.
- **Test coverage**: New `test_stream_accumulator.py` (135 lines) with thorough unit tests for the accumulator logic. Updated streaming base tests and fail-open tests to verify the aggregation behavior.
- **AgentScope instrumentation** (`_v2_middleware.py`): Significant updates to v2 middleware with proper skill instrumentation and expanded test coverage (893 lines of new tests).
- **QwenPaw instrumentation** (`patch.py`): Enhanced patch with better session tracking and turn boundary detection.
- **Multimodal upload** (`multimodal_upload_hook.py`): Improved hook with better error handling.

All changes are well-tested and follow established patterns. The streaming aggregation fix is particularly important for data quality.

LGTM ✅

---
*Automated review by "github-manager-bot"*

**Review by ralf0131 — APPROVED** [review]
### Summary
Re-review after 3 new commits since last approval (Aug 5). All changes are solid improvements:

1. **Python 3.8 source compatibility** (`bbad469e`): Mechanical replacement of `X | Y` union syntax with `Union[X, Y]` and `list[X]` with `List[X]` across all source and test files. Correct approach for maintaining 3.8 support.

2. **Finalize spans after mapping failures** (`13f0ab56`): Important defensive fix — wraps `_apply_*_response_attributes` in `try/finally` to ensure `invocation.stop()` is always called even when attribute mapping raises. Also uses `getattr()` with defaults for Usage token fields, preventing AttributeError on incomplete Usage objects. Prevents span leaks in error paths. ✅

3. **Validate Vertex AI hostname** (`a4db1558`): Security fix — replaces naive substring check (`"aiplatform.googleapis.com" in server_url`) with proper URL parsing via `urlsplit` + exact hostname match. Prevents substring confusion attacks (e.g., `aiplatform.googleapis.com.evil.example` or query parameter injection). Comprehensive test coverage including negative cases. ✅

No new issues found. LGTM.

---
*Automated review by github-manager-bot*

File: instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/client_info.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        server_url = getattr(config, "server_url", "")
+        if server_url:
+            server_address = server_url
+            server_url_string = str(server_url)
+            if "://" not in server_url_string:
+                server_url_string = "//" + server_url_string
+            server_hostname = urlsplit(server_url_string).hostname
+            if server_hostname and (
+                server_hostname == "aiplatform.googleapis.com"
+                or server_hostname.endswith("-aiplatform.googleapis.com")
```
[the remark this answers is not in the document]
Comment (sipercai) [reply instrumentation-loongsuite/loongsuite-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/client_info.py:50]: Thanks for the report. This is a false positive in this instrumentation context. The parsed hostname is used only to classify telemetry metadata (`gen_ai.provider.name` and `server.address`); it does not validate, authorize, redirect, or issue network requests. Therefore, the suggested additional suffix match is not necessary here.
